### PR TITLE
MICS-8539 Improve InstanceContext caching

### PR DESCRIPTION
**Pull request by ndussart**
Don't cache failed promise

### CODE REVIEW

**Comment by dduponchel** [comment]
I would keep the pending promise inside the cache: if not you will send a flood of requests every 10 minutes. Instead of let's say 100 concurrent requests on the plugin waiting on 1 pending promise / X http calls, the 100 requests would wait on 100 promises / 100 * X http calls, each result overriding the previous one.

**Comment by ndussart** [comment]
I made the update, is it better ?

File: src/mediarithmics/plugins/ad-renderer/base/AdRendererBasePlugin.ts [hunk 1]
```diff
@@ -101,6 +101,17 @@ export abstract class AdRendererBasePlugin<T extends AdRendererBaseInstanceConte
     instanceContext: T
   ): Promise<AdRendererPluginResponse>;
 
+  protected async getInstanceContext(creativeId: string, forceReload: boolean): Promise<T> {
+    if (!this.pluginCache.get(creativeId) || forceReload){
+      this.pluginCache.put(
+        creativeId,
+        this.instanceContextBuilder(creativeId, forceReload).catch(() => this.pluginCache.del(creativeId)),
```
Comment (dduponchel) [inline src/mediarithmics/plugins/ad-renderer/base/AdRendererBasePlugin.ts:108]: You need to re-throw the exception after the cache invalidation to propagate the error and not return a successful `undefined` (same for others).

File: src/mediarithmics/plugins/ad-renderer/base/AdRendererBasePlugin.ts [hunk 2]
```diff
@@ -101,6 +101,17 @@ export abstract class AdRendererBasePlugin<T extends AdRendererBaseInstanceConte
...
+        this.getInstanceContextCacheExpiration()
+      )
+    }
+    return Promise.resolve(this.pluginCache.get(creativeId))
```
Comment (dduponchel) [inline src/mediarithmics/plugins/ad-renderer/base/AdRendererBasePlugin.ts:112]: `pluginCache` already contains a promise, you can remove the extra `Promise.resolve()` (same for others).

**Comment by nbelie** [comment]
Jenkins retest this please

**Comment by ndussart** [comment]
retest this please

**Comment by dduponchel** [comment]
Jenkins retest this please

**Pull request merged**